content/Testing in Django - Basics - useful tools and good practices.md

```diff
@@ -0,0 +1,352 @@
+
+Since this blog is focused in API testing, I want to talk about [APITestCase](https://www.django-rest-framework.org/api-guide/testing/#api-test-cases), a class of Django REST framework, that is a mirror of TestCase, but it uses a different client class: [APIClient](https://www.django-rest-framework.org/api-guide/testing/#apiclient). This client extends the TestClient. That means that has the same functionalities, and adds some other, such as the `credentials` function. This function is very useful to overwrite authentication headers, for example using OAuth1, OAuth2, or any simple token authentication scheme.  
+So if you are working on a project with Django REST framework, you can change the previous example in this way:
+
+
+```python
+# your test file
+from rest_framework.test import APITestCase
+
+
+# class to define a test case of login
+class UserLoginTestCase(APITestCase):
+
+    ...
+```
+
+### Main testing functions
+The mentioned classes have a set of methods that will help you through the testing. With these functions you are able to corroborate that your software works as expected, and make test fixtures.  
+A test fixture is an environment that you can create to run your tests having consistency. Using fixtures you can make sure that certain conditions are met before the execution. For example, to have at the beginning a determined set of data in your testing database or to create needed objects. Here is a list of the main ones:
+- `setUp` and `tearDown`: These are functions to be executed before (setUp) and after (tearDown) each unit test. Very useful for fixtures.
+- `setUpClass` and `tearDownClass`: In an analogous way of setUp and tearDown, this functions are executed before (setUpClass) and after (tearDownClass) the whole test case. This means that are executed only once for a test case. Since they're used by Django to make important configurations, if you overwrite these methods in your test case class, don't forget to call the `super` implementation inside the functions.
+- `setUpTestData`: this function can be used to have a class-level atomic block to define data for the whole test case. That means that this function automatically rollbacks the changes in the database after the finalization of all the unit tests in the test case. Mainly used to load data to your test database.
+- Assert methods: Django has the set of [assert methods from unittest](https://docs.python.org/3/library/unittest.html#assert-methods), and has [another ones created for the framework](https://docs.djangoproject.com/en/3.2/topics/testing/tools/#assertions). These methods, such as `assertEqual`, `assertIsNone`, `assertTrue`, etc, can be used in the unit test, to check the different conditions that have to be met. For example, in the unit test of login, you can assert that the response has a status code of 200 because you want to make sure that if the data is sent correctly, then your application has to return the response with that code.  
```

Comment:
```suggestion
- Assert methods: Django has the set of [assert methods from unittest](https://docs.python.org/3/library/unittest.html#assert-methods), and has [another one created for the framework](https://docs.djangoproject.com/en/3.2/topics/testing/tools/#assertions). These methods, such as `assertEqual`, `assertIsNone`, `assertTrue`, etc, can be used in the unit test to check the different conditions that have to be met. For example, in the unit test of login, you can assert that the response has a status code of 200 because you want to make sure that if the data is sent correctly, then your application has to return the response with that code.  
```